auth_mellon_diagnostics.c

```diff
@@ -0,0 +1,807 @@
+    case LASSO_HTTP_METHOD_GET:           return "LASSO_HTTP_METHOD_GET";
+    case LASSO_HTTP_METHOD_POST:          return "LASSO_HTTP_METHOD_POST";
+    case LASSO_HTTP_METHOD_REDIRECT:      return "LASSO_HTTP_METHOD_REDIRECT";
+    case LASSO_HTTP_METHOD_SOAP:          return "LASSO_HTTP_METHOD_SOAP";
+    case LASSO_HTTP_METHOD_ARTIFACT_GET:  return "LASSO_HTTP_METHOD_ARTIFACT_GET";
+    case LASSO_HTTP_METHOD_ARTIFACT_POST: return "LASSO_HTTP_METHOD_ARTIFACT_POST";
+    case LASSO_HTTP_METHOD_PAOS:          return "LASSO_HTTP_METHOD_PAOS";
+    default:                              return "unknown";
+    }
+}
+
+const char *
+am_diag_httpd_error_level_str(request_rec *r, int level)
+{
+    switch(level) {
+    case APLOG_EMERG:   return "APLOG_EMERG";
+    case APLOG_ALERT:   return "APLOG_ALERT";
+    case APLOG_CRIT:    return "APLOG_CRIT";
+    case APLOG_ERR:     return "APLOG_ERR";
+    case APLOG_WARNING: return "APLOG_WARNING";
+    case APLOG_NOTICE:  return "APLOG_NOTICE";
+    case APLOG_INFO:    return "APLOG_INFO";
+    case APLOG_DEBUG:   return "APLOG_DEBUG";
+    case APLOG_TRACE1:  return "APLOG_TRACE1";
+    case APLOG_TRACE2:  return "APLOG_TRACE2";
+    case APLOG_TRACE3:  return "APLOG_TRACE3";
+    case APLOG_TRACE4:  return "APLOG_TRACE4";
+    case APLOG_TRACE5:  return "APLOG_TRACE5";
+    case APLOG_TRACE6:  return "APLOG_TRACE6";
+    case APLOG_TRACE7:  return "APLOG_TRACE7";
+    case APLOG_TRACE8:  return "APLOG_TRACE8";
+    default:
+        return apr_psprintf(r->pool, "APLOG_%d", level);
+    }
+}
+
+static void
+am_diag_initilaize_req(request_rec *r, apr_file_t *diag_fd,
+                       am_req_cfg_rec *req_cfg);
+
+apr_size_t
+am_diag_time_t_to_8601_buf(char *buf, apr_size_t buf_size, apr_time_t t)
+{
+    apr_size_t ret_size;
+    apr_time_exp_t tm;
+    //const char fmt[] = "%Y-%m-%dT%H:%M:%SZ";
+    const char fmt[] = "%FT%TZ";
+
+    apr_time_exp_gmt(&tm, t);
+    apr_strftime(buf, &ret_size, buf_size, fmt, &tm);
+    if (ret_size == 0) buf[0] = 0; /* on errror assure string is null terminated */
+    return ret_size;
+}
+
+char *
+am_diag_time_t_to_8601(request_rec *r, apr_time_t t)
+{
+    char *buf;
+
+    buf = apr_palloc(r->pool, ISO_8601_BUF_SIZE);
+    if (!buf) return NULL;
+
+    am_diag_time_t_to_8601_buf(buf, ISO_8601_BUF_SIZE, t);
+    return buf;
+}
+
+static int
+am_diag_open_log(server_rec *s, apr_pool_t *p)
+{
+
+    am_mod_cfg_rec *mod_cfg = am_get_mod_cfg(s);
+
+    if (!AM_DIAG_ENABLED(mod_cfg)) {
+        ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, s,
+                     "mellon diagnostics disabled for server %s", s->defn_name);
+        return 1;
+    } else {
+        ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, s,
+                     "mellon diagnostics enabled for server %s"
+                     " diag_filename=%s", s->defn_name, mod_cfg->diag_filename);
+    }
+
+    if (!mod_cfg->diag_filename || mod_cfg->diag_fd)
```

Comment:
%s with a NULL pointer will emit the string "null", it won't cause a segfault, it can be useful to know if the value is null or not. FWIW it's not possible for diag_filename to be NULL because it has a default value.